### PR TITLE
Adds a "Save" buildmode

### PR DESCRIPTION
**Pull request by Crazylemon64**
Admins are now able to save chunks of the map for later use, which are both saved on the server, and are also sent to the admin themself for later re-upload.

The "Save" buildmode has 2 options - one saves each changed variable from the norm, the other uses json data produced from `serialize` and `deserialize` to re-create objects. The former is more comprehensive, though the latter will become more accurate as more objects have those methods implemented.

:cl:Crazylemon
rscadd: Adds a "Save" buildmode for admins to preserve their creations for later. This is currently in an early stage, so don't be surprised by errors!
/:cl:

### CODE REVIEW

File: code/modules/awaymissions/maploader/writer.dm [hunk 1]
```diff
@@ -31,40 +32,63 @@ dmm_suite{
...
+		// 	cache_index = 0
+		// cache_index++
+		// var/cache_prefix = "data/map_cache/"
+		// var/cache_name = "mapcache[cache_index].txt"
+		// var/cache_path = "[cache_prefix][cache_name]"
+		// if(fexists(cache_path))
+		// 	fdel(cache_path)
+		// // Here's hoping file writes are less pricey than repeated string appends
+		// var/template_buffer_cache = file(cache_path)
+
```
Comment (Crazylemon64) [inline code/modules/awaymissions/maploader/writer.dm]: Oh yeah here's an experiment trying to use a file to speed up repeated appending.
That definitely did not work.

**Comment by Crazylemon64** [comment]
This also would possibly lay out a framework for saving the entire station between rounds. That would be insane, but it would also be possible.

**Comment by taukausanake** [comment]
Wait, so does this save maps and edited vars? Does it outsource this to a file that can be read outside of the game like in DreamMaker? As in you can edit the world in-game and save/export it and copy/paste the edited section into map proper?

**Comment by Crazylemon64** [comment]
@taukausanake The non-json version would be capable of seamlessly working with DM's map editor, if we properly set `/tmp` on all vars where it made sense. We don't, however, so it's probably going to be a TAD messy to do so, but you still can edit those maps without a hitch.

As for the json-ized version, I'm not sure. Since objects don't normally have a `map_json_data` var, I'm thinking DM would give it some straaaaange looks.

**Comment by taukausanake** [comment]
@Crazylemon64 So long as I can edit the maps in DM I am SUPER HAPPY :)

**Comment by monster860** [comment]
Are you making sure to remove objects like lighting overlays?

**Comment by Crazylemon64** [comment]
I did not do anything to handle lighting overlays - that's next on the list of things to patch up with this

**Comment by Crazylemon64** [comment]
Lighting overlays handled.

**Comment by Plotron** [comment]
I misread "Slave" gamemode. Luckily we're not back to good old America...

**Comment by Necaladun** [comment]
Holy shit. Yes.

**Comment by Crazylemon64** [comment]
Hold on, I need to make JSON mode output a map that looks legible in DM, and make the writer.dm file not be full of formatting that conflicts with what we normally use

**Comment by Crazylemon64** [comment]
Blast it it didn't work

**Comment by Crazylemon64** [comment]
DM can now read the map files in both JSON and non-JSON mode, and the map writer is now 19x faster, being capable of saving the cyberiad in 5.6 seconds, as opposed to the 80 seconds it took before.

File: code/modules/awaymissions/maploader/writer.dm [hunk 2]
```diff
@@ -47,7 +47,8 @@
 	var/turf/ne = locate(max(t1.x,t2.x),max(t1.y,t2.y),max(t1.z,t2.z)) // Outer corner
 	var/turf/sw = locate(min(t1.x,t2.x),min(t1.y,t2.y),min(t1.z,t2.z)) // Inner corner
 	var/list/templates[0]
-	var/template_buffer = ""
+	var/list/template_buffer = list()
+	var/template_buffer_text
 	var/buffer_line = ""
```
Comment (tigercat2000) [inline code/modules/awaymissions/maploader/writer.dm]: is this even used now

Comment (Crazylemon64) [inline code/modules/awaymissions/maploader/writer.dm]: Oh, no it isn't. Silly me.

**Comment by Crazylemon64** [comment]
:bomb: :fire: :zap:
:eyes:

**Comment by Fox-McCloud** [comment]
Done with this @Crazylemon64 ?

**Comment by Crazylemon64** [comment]
@Fox-McCloud I have made all changes that I deem necessary for the map writer component. More elaborate serialization support can come later.

**Pull request merged**